### ISSUE TITLE
Background worker aborts due to file locking error

### ISSUE THREAD

**Issue by storedcc** [zammad/zammad#4294]
### Infos:

* Used Zammad version: 5.2.3-11
* Installation method (source, package, ..): Kubernetes / [Helm](https://github.com/zammad/zammad-helm)
* Operating system: K3S / Ubuntu 22.04 with Longhorn as Storage Provisioner
* Database + version: postgresql / 11.14.0-debian-10-r28
* Elasticsearch version: 5.2.3-11
* Browser + version: N/A

### Expected behavior:

* File locking (flock) of /opt/zammad/lib/background_service/cli.rb should work properly

### Actual behavior:

* File locking (flock) of /opt/zammad/lib/background_service/cli.rb throws errors:
```
I, [2022-10-11T11:10:37.895152 #1-7260]  INFO -- : Using memcached as Rails cache store.
I, [2022-10-11T11:10:37.899469 #1-7260]  INFO -- : Using Redis as web socket session store.
I, [2022-10-11T11:10:42.920957 #1-7260]  INFO -- : Setting.set('models_searchable', ["Chat::Session", "KnowledgeBase::Answer::Translation", "Organization", "Ticket", "User"])
bundler: failed to load command: script/background-worker.rb (script/background-worker.rb)
/opt/zammad/lib/background_services/cli.rb:45:in `flock': Bad file descriptor @ rb_file_flock - /opt/zammad/lib/background_services/cli.rb (Errno::EBADF)
 from /opt/zammad/lib/background_services/cli.rb:45:in `lock'
 from /opt/zammad/lib/background_services/cli.rb:14:in `start'
 from /usr/local/bundle/gems/thor-1.2.1/lib/thor/command.rb:27:in `run'
 from /usr/local/bundle/gems/thor-1.2.1/lib/thor/invocation.rb:127:in `invoke_command'
 from /usr/local/bundle/gems/thor-1.2.1/lib/thor.rb:392:in `dispatch'
 from /usr/local/bundle/gems/thor-1.2.1/lib/thor/base.rb:485:in `start'
 from script/background-worker.rb:16:in `<top (required)>'
 from /usr/local/lib/ruby/3.0.0/bundler/cli/exec.rb:58:in `load'
 from /usr/local/lib/ruby/3.0.0/bundler/cli/exec.rb:58:in `kernel_load'
 from /usr/local/lib/ruby/3.0.0/bundler/cli/exec.rb:23:in `run'
 from /usr/local/lib/ruby/3.0.0/bundler/cli.rb:479:in `exec'
 from /usr/local/lib/ruby/3.0.0/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
 from /usr/local/lib/ruby/3.0.0/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
 from /usr/local/lib/ruby/3.0.0/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
 from /usr/local/lib/ruby/3.0.0/bundler/cli.rb:31:in `dispatch'
 from /usr/local/lib/ruby/3.0.0/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
 from /usr/local/lib/ruby/3.0.0/bundler/cli.rb:25:in `start'
 from /usr/local/lib/ruby/gems/3.0.0/gems/bundler-2.2.33/libexec/bundle:49:in `block in <top (required)>'
 from /usr/local/lib/ruby/3.0.0/bundler/friendly_errors.rb:103:in `with_friendly_errors'
 from /usr/local/lib/ruby/gems/3.0.0/gems/bundler-2.2.33/libexec/bundle:37:in `<top (required)>'
 from /usr/local/bin/bundle:23:in `load'
 from /usr/local/bin/bundle:23:in `<main>'
 ```

The problematic code line can be found [here](https://github.com/zammad/zammad/blob/develop/lib/background_services/cli.rb#L44)

This can be "fixed" by either specifying a different lock file position (For example `@lock_file = File.open('/tmp/background.lock', 'r')`)
Or by setting the File.open parameter to write mode: `@lock_file = File.open(__FILE__, 'w')`. 
Though I am more than aware that this isn't a fix, but a hacky workaround I'd like to avoid. Maybe this can be fixed? If there is further testing or more information needed, please let me know.

### Steps to reproduce the behavior:

* Deploy Zammad with Helm Chart on specified (very common) Kubernetes Cluster / storage combination.

Yes I'm sure this is a bug and no feature request or a general question.

**Comment by mgruner** [comment]
@storedcc thanks a lot for your report. Very interesting!

The purpose of this code is to prevent multiple parallel invocations of the background worker. For this, it locks itself by opening the `cli.rb` for reading and then places a `flock` on it. It seems that `@lock_file = File.open(__FILE__, 'r')` does not produce a valid file handle, or that the subsequent `flock` call produces the error. That seems strange.

What kind of file system is this? Could it be a network/cluster file system that does not support `flock`? In that case, a workaround could be placing a separate lock file in Zammad's `tmp` directory. Or is that on the same file system and would suffer the same issue?

**Comment by mgruner** [comment]
See also https://github.com/rubygems/rubygems/issues/806

**Comment by MrGeneration** [comment]
> What kind of file system is this?

![image](https://user-images.githubusercontent.com/6549061/195552026-37b4045b-45f9-4764-8963-b054e2c3a4b0.png)


I'd guess our docker-compose. If not, most likely out of our scope as we can't support any wild system out there.

**Comment by storedcc** [comment]
> What kind of file system is this?

As far as I can tell, the official Helm deployment uses the docker-compose variants of the Zammad Image.

Rancher Longhorn itself uses ext4 as the backing file system, so nothing out of the ordinary. In my opinion, a flock, as it is temporary, should go to a file in /tmp (Or anywhere outside the mounted /opt/zammad/) to provide maximum compatibility across all filesystems and to avoid issues as this, if it is caused by flocks not being supported for some reason.

As /tmp should in general never be mounted and would always be on ephemeral storage which should support flock without issues, this might be an option and could be changed in the Zammad code.

**Comment by storedcc** [comment]
> If not, most likely out of our scope as we can't support any wild system out there.

I totally agree that supporting any wild filesystem is way out of scope, though in this case the usage of Longhorn / ext4 should be a quite common way of doing things given the fact we are talking about the official Helm chart provided by Zammad. 

My suggestion as written above would be to utilize an ephemeral partition within the container to "store" the flocks. This should prevent any further issues which might pop up regarding that topic. I also don't see the need to place the flock onto the ruby file which is executed (other than the convenience that __FILE__ provides a neat pre-defined variable, please correct me here if there's a valid reason, I might be incorrect)

**Comment by danat-cc** [comment]
I have the same issue. The latest version of zammad I was able to get running with helm was: helm chart 6.6.0 and tag 5.1.1-57. It seems with the move to 5.2 the code involving flock was introduced. 

I must agree with this sentiment of the ticket author - this is a major issue for us. We use k8s, NFS 4.2 for volumes (ext4 as underlying fs on the nfs server) and we are seeing this issue as well. I'm frankly a bit skeptical that this is an issue on bare ext4. @storedcc Is it possible that you're Longhorn setup is also using an NFS provisioner that is then available over multiple hosts? I haven't looked into it aside from looking it up just now but it's "cloud native distributed block storage" so it's conceivable that NFS is used for assigning volumes to pods. 

flock and nfs3/4 can be tricky and I've seen the same behavior with checkmk unfortunately. to get that to work required a hacky emptydir mount for the main server data which gets restored from backup from nfs share during init (and recurring backups as cronjob). Very ugly... 

From [what google tells me](https://stackoverflow.com/questions/29611352/what-is-the-difference-between-locking-with-fcntl-and-flock) fcntl and nfs get along better... any chance of getting support for nfs backend? in containerized workloads it is very common.

**Comment by MrGeneration** [comment]
I'd like to point this out just so that we all have a -current as is- standing point:

![image](https://user-images.githubusercontent.com/6549061/202028582-1c512900-7c1e-4e29-9718-5d56d09a5af2.png)
![image](https://user-images.githubusercontent.com/6549061/202028600-d9bc770c-ee9d-4d50-9045-46d647c95f7f.png)

**Comment by mgruner** [comment]
This was caused by a mechanism that tried to prevent multiple executions of the background worker on the same machine (through file locking).

We decided to remove this, mainly because it does not work properly in clusters anyway, because the previous scheduler also didn't do that check (if started in foreground mode) and because it's the job of tools like systemd to prevent parallel execution. So this `flock` is now gone, which avoids problems with file systems and POSIX vs. non-POSIX platforms. Back port to Zammad 5.2 should land soon.
